<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li></ul></div>

# Защита персональных данных клиентов

## Загрузка данных

In [2]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

import numpy as np

In [3]:
data = pd.read_csv('/datasets/insurance.csv')
data.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [6]:
data.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

In [7]:
data.corr()

Пол   Возраст  Зарплата  Члены семьи  \
Пол                1.000000  0.002074  0.014910    -0.008991   
Возраст            0.002074  1.000000 -0.019093    -0.006692   
Зарплата           0.014910 -0.019093  1.000000    -0.030296   
Члены семьи       -0.008991 -0.006692 -0.030296     1.000000   
Страховые выплаты  0.010140  0.651030 -0.014963    -0.036290   

                   Страховые выплаты  
Пол                         0.010140  
Возраст                     0.651030  
Зарплата                   -0.014963  
Члены семьи                -0.036290  
Страховые выплаты           1.000000

In [4]:
data.columns = ['f1', 'f2', 'f3', 'f4', 't']

In [5]:
features = data.drop('t', axis=1).values
features

array([[1.00e+00, 4.10e+01, 4.96e+04, 1.00e+00],
       [0.00e+00, 4.60e+01, 3.80e+04, 1.00e+00],
       [0.00e+00, 2.90e+01, 2.10e+04, 0.00e+00],
       ...,
       [0.00e+00, 2.00e+01, 3.39e+04, 2.00e+00],
       [1.00e+00, 2.20e+01, 3.27e+04, 3.00e+00],
       [1.00e+00, 2.80e+01, 4.06e+04, 1.00e+00]])

In [6]:
target = data['t'].values
target

array([0, 1, 0, ..., 0, 0, 0])

In [7]:
features_train, features_valid, target_train, target_valid  = \
    train_test_split(features, target, test_size=0.25, random_state=12345)

**Вывод**

В исходных данных 5000 строк и 5 колонок, пропусков нет.
Переименовала колонки в незначащие названия.
Выделила матрицу признаков и вектор целевого признака и разделила на тестовую и обучающую выборку.

В таблице данных компании соблюден почти идеальный баланс между мужчинами и женщинами. Средний возраст застрахованного составляет около 31 года, при этом медиана составляет 30. Границы возраста застрахованных приводят к выводу, что компания не занимается страхованием лиц, чей возраст составляет менее 18 и старше 65 лет. Также интересно взглянуть на уровень заработной платы застрахованных - медиана составляет чуть более 40 000 рублей. В целом, в компании скорее представлен средний экономический класс, чем слишком бедные или слишком богатые граждане.

Возраст и количество страховых выплат клиенту за последние 5 лет имеют корреляцию 0.651. Вероятно, для компании есть взаимосвязь: чем старше застрахованное лицо, тем более вероятно он потребует страховку чаще, следовательно, несёт в себе больше рисков для предприятия.

## Умножение матриц

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Вопрос:** Изменится ли качество линейной регрессии при умножении признаков на обратимую матрицу?

**Ответ:** Не изменится.

**Обоснование:**

Предсказания в исходной модели:

$$
a = Xw
$$

где

$$
w = (X^T X)^{-1} X^T y
$$

Предсказания в преобразованной модели:

$$
a = X P v
$$

где

$$
v = ((X P)^T (X P))^{-1} (X P)^T y
$$

Преобразуем:

$$
v = (P^T P)^{-1} P^T (X^T X)^{-1} X^T y
$$

$$
v = P^{-1} w
$$

Подставим полученное в предсказания в преобразованной модели:

$$
a = X P P^{-1} w
$$

И получим предсказания в исходной модели:

$$
a = Xw
$$

Таким образом, при умножении матрицы признаков на обратимую матрицу результат предсказаний линейной регрессии не меняется.

## Алгоритм преобразования

**Алгоритм**

Для "обезличивания" персональных данных можно умножить матрицу признаков на случайную обратимую матрицу размером 4х4.

**Обоснование**

Такое преобразование не изменит качество линейной регрессии, т.к. оно будет учтено при расчете вектора коэффициентов модели - матрица, обратная случайной матрице, взятой для преобразования, будет умножена на вектор коэффициентов (какой бы он был при обучении модели без преобразования). И в моделе будут взяты полученные в итоге этого произведения коэффициенты.

## Проверка алгоритма

In [8]:
def get_r2(features_train, features_valid, target_train, target_valid):
    model = LinearRegression()
    model.fit(features_train, target_train)
    return r2_score(target_valid, model.predict(features_valid))

In [9]:
get_r2(features_train, features_valid, target_train, target_valid)

0.435227571270266

In [10]:
np.random.seed(12345)
p = np.random.sample((4, 4))

Проверили, что полученная случайная матрица - обратимая.

In [11]:
np.linalg.inv(p)

array([[ 2.1336149 ,  0.5051926 , -1.71113385,  1.33832486],
       [-3.76534462, -2.61352053,  6.72316006, -6.07193701],
       [ 0.97426298,  2.9928115 , -3.49875049,  1.84220045],
       [ 0.13998298, -0.94454066,  0.52375812,  1.652661  ]])

In [12]:
get_r2(features_train @ p, features_valid @ p, target_train, target_valid)

0.4352275712706706

В итоге получим такой "обезличенный" датасет:

In [13]:
data = pd.DataFrame(data = np.concatenate((features @ p, np.array([target]).T), axis=1), columns = data.columns)
data.head()

f1            f2            f3            f4    t
0  37169.983952  32441.905747  37126.682258  47708.455342  0.0
1  28484.575966  24863.156593  28457.829124  36560.358528  1.0
2  15743.503414  13742.238084  15729.981916  20206.383593  0.0
3  31241.345787  27266.582942  31200.559759  40101.520436  0.0
4  19563.289157  17075.165256  19542.546847  25108.599330  0.0

**Вывод**

Эмпирическим путем получили те же выводы, которые были сделаны выше на основе теории.

Таким образом, в случае домножения матрицы признаков на случайную обратимую матрицу можно обеспечить защиту данных, не особо потеряв в качестве модели (незначительное отличие вызвано особенностью преобразования матриц и чисел с плавающей точкой, это норма).